compiler/lib/src/main/scala/codegen/CppWriter/StructCppWriter.scala

```diff
@@ -0,0 +1,775 @@
+    val strTypes = members.map((_, t) => t match {
+      case t: Type.String => Some(t)
+      case _ => None
+    }).filter(_.isDefined).map(_.get).toList
+    strTypes match {
+      case Nil => Nil
+      case l => strCppWriter.write(l)
+    }
+  }
+
+  private def getConstantMembers: List[CppDoc.Class.Member] =
+    List(
+      CppDoc.Class.Member.Lines(
+        CppDoc.Lines(
+          CppDocHppWriter.writeAccessTag("public") ++
+            CppDocWriter.writeBannerComment("Constants") ++
+            addBlankPrefix(
+              wrapInEnum(
+                List(
+                  lines("//! The size of the serial representation"),
+                  lines("SERIALIZED_SIZE ="),
+                  lines(memberList.map((n, tn) =>
+                    s.getSerializedSizeExpr(members(n), tn) + (
+                      if sizes.contains(n) then s" * ${sizes(n)}"
+                      else ""
+                      )).mkString(" +\n")).map(indentIn),
+                ).flatten
+              )
+            )
+        )
+      )
+    )
+
+  private def getTypeMembers: List[CppDoc.Class.Member] =
+    if sizes.isEmpty then Nil
+    else
+      // Write types for array members to simplify C++ array reference syntax
+      List(
+        CppDoc.Class.Member.Lines(
+          CppDoc.Lines(
+            List(
+              CppDocHppWriter.writeAccessTag("public"),
+              CppDocWriter.writeBannerComment("Types"),
+              Line.blank :: lines("//! The array member types"),
+              memberList.filter((n, _) => sizes.contains(n)).map((n, tn) =>
+                line(s"typedef $tn ${getArrayTypeName(n)}[${sizes(n)}];")
+              )
+            ).flatten
+          )
+        )
+      )
+
+  private def getConstructorMembers: List[CppDoc.Class.Member] = {
+    val defaultValues = getDefaultValues
+    // Only write this constructor if the struct contains an array
+    val scalarConstructor =
+      if sizes.isEmpty then None
+      else Some(
+        CppDoc.Class.Member.Constructor(
+          CppDoc.Class.Constructor(
+            Some("Member constructor (scalar values for arrays)"),
+            memberList.map(writeMemberAsParamScalar),
+            writeInitializerList(n => n),
+            writeArraySetters(n => n)
+          )
+        )
+      )
+
+    List(
+      CppDoc.Class.Member.Lines(
+        CppDoc.Lines(
+          CppDocHppWriter.writeAccessTag("public")
+        )
+      ),
+      CppDoc.Class.Member.Lines(
+        CppDoc.Lines(
+          CppDocWriter.writeBannerComment("Constructors"),
+          CppDoc.Lines.Both
+        )
+      ),
+      CppDoc.Class.Member.Constructor(
+        CppDoc.Class.Constructor(
+          Some("Constructor (default value)"),
+          Nil,
+          "Serializable()" :: nonArrayMemberNames.map(n => {
+            val defVal = defaultValues(n)
+            // Parse non-literal default values, e.g. T() for abstract type T
```

Comment:
Why do we need this special-case handling? If we need it, let's not parse string output. Let's match against FPP semantic concepts, e.g., types.

Reply:
Looks good!